processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java

```diff
@@ -1371,6 +1361,77 @@ public Grouper.BufferComparator bufferComparatorWithAggregators(
       );
     }
 
+    @Override
+    public ObjectMapper decorateObjectMapper(ObjectMapper spillMapper)
+    {
+
+      final JsonDeserializer<RowBasedKey> deserializer = new JsonDeserializer<RowBasedKey>()
+      {
+        @Override
+        public RowBasedKey deserialize(
+            JsonParser jp,
+            DeserializationContext deserializationContext
+        ) throws IOException
+        {
+          if (!jp.isExpectedStartArrayToken()) {
+            throw DruidException.defensive("Expected array start token, received [%s]", jp.getCurrentToken());
+          }
+          jp.nextToken();
+
+          final ObjectCodec codec = jp.getCodec();
+          final int timestampAdjustment = includeTimestamp ? 1 : 0;
+          final int dimsToRead = timestampAdjustment + serdeHelpers.length;
+          int dimsReadSoFar = 0;
+          final Object[] objects = new Object[dimsToRead];
+
+          while (jp.currentToken() != JsonToken.END_ARRAY) {
+            if (dimsReadSoFar >= dimsToRead) {
+              throw DruidException.defensive("More dimensions encountered than expected [%d]", dimsToRead);
+            }
+
+            if (includeTimestamp && dimsReadSoFar == 0) {
+              // Read the timestamp
+              objects[dimsReadSoFar] = codec.readValue(jp, Long.class);
+            } else {
+              DruidException.conditionalDefensive(
+                  dimsReadSoFar - timestampAdjustment < serdeHelpers.length,
+                  "Insufficient serde helpers present"
+              );
+              // Read the dimension
+              if (serdeHelpers[dimsReadSoFar - timestampAdjustment].getComplexClazz() == null) {
+                objects[dimsReadSoFar] = codec.readValue(jp, Object.class);
+                if (objects[dimsReadSoFar] instanceof Integer) {
+                  objects[dimsReadSoFar] = ((Integer) objects[dimsReadSoFar]).longValue();
+                } else if (objects[dimsReadSoFar] instanceof Double) {
+                  objects[dimsReadSoFar] = ((Double) objects[dimsReadSoFar]).floatValue();
+                }
```

Comment:
could serde helpers just tell this to use `Float.class` and `Long.class` instead of generically doing `Object.class` and then checking it? (e.g. make key serde helpers return non-null classes instead of only complex types being non-null)

Reply:
I have modified this. I had a doubt about the original flow. In the case of Double dimensions, we would deserialize them as double and cast them as float right, which seems incorrect.